velox/functions/prestosql/aggregates/ReservoirSampleAccumulator.h

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <random>
+#include "velox/common/base/BitUtil.h"
+#include "velox/functions/lib/aggregates/ValueList.h"
+#include "velox/type/FloatingPointUtil.h"
+#include "velox/type/StringView.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::aggregate::prestosql {
+
+struct ReservoirSampleAccumulator {
+  VectorPtr samples;
+  VectorPtr initialSamples;
+  int64_t initialSeenCount = 0;
+  int64_t processedCount = 0;
+  int32_t maxSampleSize = -1;
+
+  explicit ReservoirSampleAccumulator(memory::MemoryPool* pool) : pool_(pool) {}
+
+  bool isInitialized() const {
+    return maxSampleSize >= 0;
+  }
+
+  int32_t initialSampleCount() const {
+    return initialSamples ? initialSamples->size() : 0;
+  }
+
+  vector_size_t sampleCount() const {
+    return samples ? samples->size() : 0;
+  }
+
+  void addValueToReservoir(
+      const DecodedVector& decodedVector,
+      vector_size_t row,
+      const TypePtr& type) {
+    if (processedCount <= maxSampleSize) {
+      appendValue(decodedVector, row, type);
+      return;
+    }
+
+    int64_t replaceIndex = -1;
+    if (!shouldReplaceSample(decodedVector, row, replaceIndex)) {
+      return;
+    }
+
+    replaceValueAt(decodedVector, row, replaceIndex);
+  }
+
+  bool shouldReplaceSample(
+      const DecodedVector& decodedVector,
+      vector_size_t row,
+      int64_t& replaceIndex) const {
+    std::mt19937_64 rng(static_cast<uint64_t>(processedCount));
```

Comment:
Do you need to create a new rng on every call to determine if you should replace a sample ? This construction is not cheap, You can instead create the uniform distribution each time and just reuse the rng ? Another way to do it is to just have an an rng between 0 to 1 and scale it each time with processedCount (?).